### PR TITLE
VSL-121: Deposit Token Flow

### PR DESCRIPTION
**Pull request by KsenijaZivkovic**
## Description

Modal for depositing tokens was added
Works on clicking all deposit buttons in the app (on the home screen, and on the tokens page)
A button for deposit on Tokens page by selecting vault is only enabled if the user has that vault in their wallet.
Send button on Tokens page is disabled if the vault is empty

## Demo / Test Result

Screen Shot 2022-10-07 at 12 38 20 AM

Screen Shot 2022-10-07 at 12 38 46 AM

Screen Shot 2022-10-12 at 11 43 40 AM

In case of error, when user tries to send all the FLOW tokens to the treasury:
Screen Shot 2022-10-07 at 1 17 16 PM
Improvement created - https://linear.app/dappercollectives/issue/VSL-256

Tooltip:
Screen Shot 2022-10-07 at 1 18 09 PM
Improvement created - https://linear.app/dappercollectives/issue/VSL-254

### CODE REVIEW

**Comment by linxiao-li** [comment]
1. Suggest renaming the directory and files to reflect the update 
`src/components/SendToken` (only send token flow)-> `src/components/TransferToken` (both send and deposit flow)
`src/components/SendToken/SendTokenContext` -> `src/components/SendToken/TransferTokenContext`

2. DepositToken.js and SendToken.js is the same except for the safedata and userAddress properties, and those two can be accessed anywhere in the child component using the web3 context, suggest combine them into one `TransferTokenContext`

3. Instead of doing a bunch of conditional rendering in the child components (i.e. ` if(transactionType === TRANSACTION_TYPE.SEND){...}`), can we try to abstract the interface for the child component and pass executables and properties? That can help reducing the complexity of the code

**Review by linxiao-li — COMMENTED** [review]
A lot of changes are prettier quote changes, @mannynotfound can you create a pr that takes care of the quote changes all together? We can merge that one first. That will help making this pr cleaner

**Comment by amyliu6** [comment]
@KsenijaZivkovic  re ```Send button on Tokens page is hidden if the vault is empty``` , wasn't aware that we need to hide 'send' button when vault is empty, could you help clarify if that's a new design change request from Alaa?

**Comment by KsenijaZivkovic** [comment]
> @KsenijaZivkovic re `Send button on Tokens page is hidden if the vault is empty` , wasn't aware that we need to hide 'send' button when vault is empty, could you help clarify if that's a new design change request from Alaa?

@amyliu6 I have proposed this change, because I feel that it shouldn't be possible to send tokens from the empty vault. Even if the user opens the vault they wouldn't be able to change the amount and BE also prevents the users to propose 0 tokens to account. Let me know your thoughts, I can easily revert it!

**Comment by amyliu6** [comment]
> > @KsenijaZivkovic re `Send button on Tokens page is hidden if the vault is empty` , wasn't aware that we need to hide 'send' button when vault is empty, could you help clarify if that's a new design change request from Alaa?
> 
> @amyliu6 I have proposed this change, because I feel that it shouldn't be possible to send tokens from the empty vault. Even if the user opens the vault they wouldn't be able to change the amount and BE also prevents the users to propose 0 tokens to account. Let me know your thoughts, I can easily revert it!

That true. As there were design exploration to notify user about empty vault in "Send flow" (for future release), hence want to double check here, we should only update design after getting their approval.

**Comment by KsenijaZivkovic** [comment]
> A lot of changes are prettier quote changes, @mannynotfound can you create a pr that takes care of the quote changes all together? We can merge that one first. That will help making this pr cleaner

@linxiao-li PR created: https://github.com/DapperCollectives/VESSEL/pull/130

File: packages/client/src/components/SafeHome.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/SafeHome.js:21]: safeData can be accessed if we pass through the address, I will create a commit soon so you can take a look

File: packages/client/src/components/SafeNFTs.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/SafeNFTs.js]: Let me know if I changed it correctly!

File: packages/client/src/components/SendTokens/steps/DepositTokenForm.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/SendTokens/steps/DepositTokenForm.js]: > > > @KsenijaZivkovic re `Send button on Tokens page is hidden if the vault is empty` , wasn't aware that we need to hide 'send' button when vault is empty, could you help clarify if that's a new design change request from Alaa?
> > 
> > 
> > @amyliu6 I have proposed this change, because I feel that it shouldn't be possible to send tokens from the empty vault. Even if the user opens the vault they wouldn't be able to change the amount and BE also prevents the users to propose 0 tokens to account. Let me know your thoughts, I can easily revert it!
> 
> That true. As there were design exploration to notify user about empty vault in "Send flow" (for future release), hence want to double check here, we should only update design after getting their approval.

Updated as per design, to disable buttons instead!

File: packages/client/src/components/SendTokens/steps/SendTokenForm.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/SendTokens/steps/SendTokenForm.js]: In SendTokenForm I am fetching the balances now from the Web3 context
But in deposit form the balances of the user's vault should be fetched each time the deposit token is opened, to make sure the balances are correct (it can change every second, and it is not controlled by Vessel)

File: packages/client/src/components/SendTokens/steps/DepositTokenForm.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/SendTokens/steps/DepositTokenForm.js]: > Can you help me understand the purpose of this effect? Looks like it will only be called on the deposit token modal open since vaults won't change during the depositing flow?

It was the only way to sync treasury vaults and user vaults -> to fetch the vaults from the treasury first and then to update the vaults of the user that can be used to deposit tokens.

File: packages/client/src/components/SendTokens/components/ButtonGroup.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/SendTokens/components/ButtonGroup.js:52]: I've extracted the function for transferring tokens out of button group

**Comment by KsenijaZivkovic** [comment]
> > > @KsenijaZivkovic re `Send button on Tokens page is hidden if the vault is empty` , wasn't aware that we need to hide 'send' button when vault is empty, could you help clarify if that's a new design change request from Alaa?
> > 
> > 
> > @amyliu6 I have proposed this change, because I feel that it shouldn't be possible to send tokens from the empty vault. Even if the user opens the vault they wouldn't be able to change the amount and BE also prevents the users to propose 0 tokens to account. Let me know your thoughts, I can easily revert it!
> 
> That true. As there were design exploration to notify user about empty vault in "Send flow" (for future release), hence want to double check here, we should only update design after getting their approval.

Updated according to the designs, buttons are disabled instead of hidden!

File: packages/client/src/hooks/useAccount.js
Comment (amyliu6) [inline packages/client/src/hooks/useAccount.js:86]: do we need `getUserVaults` func? can we use `getUserBalances` and check if `balance is > 0` to know whether vault exist?

File: packages/client/src/hooks/useAccount.js
Comment (amyliu6) [inline packages/client/src/hooks/useAccount.js:10]: if it's now entire list maybe this line can be removed and directly use it inside method.

File: packages/client/src/hooks/useAccount.js
Comment (KsenijaZivkovic) [reply packages/client/src/hooks/useAccount.js:86]: Vault can exist with 0 balance, and it's difficult to differentiate if the vault exists or not if both return 0 as a result - but in our use case it's okay to do it in a way you proposed!

File: packages/client/src/components/TransferToken/components/ButtonGroup.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/TransferToken/components/ButtonGroup.js:31]: Address in this case is the sender address, so in case of sending tokens - it's a treasury address, and in case of depositing -> it's the user's address. I will rename this to sender, so it will make more sense!

File: packages/client/src/flow/getAccountBalanceByContractName.tx.js
Comment (KsenijaZivkovic) [reply packages/client/src/flow/getAccountBalanceByContractName.tx.js:8]: I can revert it to first version if we go with the vault > 0 for depositing!

File: packages/client/src/hooks/useAccount.js
Comment (amyliu6) [inline packages/client/src/hooks/useAccount.js:83]: `result.filter((coin) => coin.balance)` for readability

File: packages/client/src/hooks/useAccount.js
Comment (KsenijaZivkovic) [reply packages/client/src/hooks/useAccount.js:10]: I would leave it like this if that is okay, because this variable is used in a couple of places so might be easier to read the code

File: packages/client/src/hooks/useAccount.js
Comment (KsenijaZivkovic) [reply packages/client/src/hooks/useAccount.js:86]: On page where we see treasury vaults -> we need the information if we should enable/disable the button for depositing tokens, there we don't need info about balance, but rather if the vault exists or not on the user side -> so that way we avoid more requests than needed (the number of requests doubles if we call it on that page).

Comment (amyliu6) [reply packages/client/src/hooks/useAccount.js:86]: make sense. thanks for explaining to me!

**Comment by amyliu6** [comment]
@KsenijaZivkovic  has mobile web UI been implemented in this PR? if so could you please attach screenshot for that too?

**Comment by KsenijaZivkovic** [comment]
> @KsenijaZivkovic has mobile web UI been implemented in this PR? if so could you please attach screenshot for that too?

No, currently both send/deposit modals aren't mobile responsive

File: packages/client/src/components/TransferToken/steps/index.js
Comment (linxiao-li) [inline packages/client/src/components/TransferToken/steps/index.js:39]:
```suggestion
{currentStep === 1 && <TransferTokenConfirmation/>}
{currentStep ===0 && isSendTransaction &&<SendTokenForm/>}
{currentStep === 0 && !isSendTransaction &&<DepositTokenForm/>}
```

File: packages/client/src/components/TransferToken/components/TransactionDetails.js
Comment (mannynotfound) [inline packages/client/src/components/TransferToken/components/TransactionDetails.js:48]: since theres only 2 conditions this would probably be clearer as an if/else statement - unless there is other transaction types coming in which case switch is probably better

File: packages/client/src/components/TransferToken/steps/DepositTokenForm.js
Comment (mannynotfound) [inline packages/client/src/components/TransferToken/steps/DepositTokenForm.js:53]:
```suggestion
   // eslint-disable-next-line react-hooks/exhaustive-deps
```

we should probably get out of the habit of abusing this eslint ignore though, apparently its not _good code_ https://dev.to/wkrueger/never-ignore-the-exhaustive-deps-rule-2ap8

File: packages/client/src/hooks/useAccount.js
Comment (mannynotfound) [inline packages/client/src/hooks/useAccount.js:91]:
```suggestion
        coin.exists = await getVaultExists(contractName, vaultPath, address);
```

File: packages/client/src/modals/Transaction/BannerInfo.js
Comment (mannynotfound) [inline packages/client/src/modals/Transaction/BannerInfo.js:25]: this should conform to `MetadataViews.cdc` and use `thumbnail` like in this PR: https://github.com/DapperCollectives/VESSEL/pull/131

**Review by mannynotfound — COMMENTED** [review]
looks pretty good to me! 

metadata image field needs to be changed + a few suggestions

File: packages/client/src/components/SendTokens/steps/DepositTokenForm.js
Comment (linxiao-li) [reply packages/client/src/components/SendTokens/steps/DepositTokenForm.js]: I guess I'm missing the point of having two separate useEffect hooks, why not combine with the hook above?

File: packages/client/src/components/TransferToken/components/TransactionDetails.js
Comment (linxiao-li) [inline packages/client/src/components/TransferToken/components/TransactionDetails.js:30]:
```suggestion
  const safeName = web3?.treasuries?.[safeAddress]?.name;
```

File: packages/client/src/components/SendTokens/steps/SendTokenForm.js
Comment (linxiao-li) [reply packages/client/src/components/SendTokens/steps/SendTokenForm.js]: What I mean is more about where to put the data after fetching (either from web3 context or from api call for user's vaults), right now we fetch data on `SendTokenForm` and `DepositTokenForm` then pass them down through the property chain via property `coinBalances`, what I'm suggesting is putting the coinBalances data into the context, and have cointypedropdown reading from the context. Let me know if it makes more sense.

File: packages/client/src/components/TransferToken/components/CoinTypeDropDown.js
Comment (linxiao-li) [inline packages/client/src/components/TransferToken/components/CoinTypeDropDown.js:9]: What's the intention for the change here? Do we want to show an empty string or 0 when the balances[key] is undefined?

File: packages/client/src/components/SafeNFTs.js
Comment (linxiao-li) [reply packages/client/src/components/SafeNFTs.js]: All good!

File: packages/client/src/components/SendTokens/steps/SendTokenForm.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/SendTokens/steps/SendTokenForm.js]: Ohhh right, makes sense! I will put it inside the context and fetch it instead of passing through property. Thank you!

File: packages/client/src/components/SendTokens/steps/DepositTokenForm.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/SendTokens/steps/DepositTokenForm.js]: I'll try again and change it if it works in one useEffect hook.

File: packages/client/src/components/TransferToken/components/TransactionDetails.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/TransferToken/components/TransactionDetails.js:48]: No, it's only two of them, I'll update it!

File: packages/client/src/components/SendTokens/steps/DepositTokenForm.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/SendTokens/steps/DepositTokenForm.js]: It causes infinite loops when merged into one useEffect. The use case that this is solving is when user opens the page for the first time and local storage is empty, when vaults are loaded to update the user balance

File: packages/client/src/components/TransferToken/TransferTokensContext.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/TransferToken/TransferTokensContext.js:24]: Makes sense to put it inside contexts, even though it's only used here

File: packages/client/src/modals/Transaction/BannerInfo.js
Comment (KsenijaZivkovic) [reply packages/client/src/modals/Transaction/BannerInfo.js:25]: It doesn't work if I change it to thumbnail, the banner doesn't show an image.
If we make it work for ExampleNFT, how will it behave on testnet for all the NFTs that are deployed there?
Examples can be found here: https://nft-catalog.vercel.app/catalog/testnet/Momentables

File: packages/client/src/components/TransferToken/components/CoinTypeDropDown.js
Comment (KsenijaZivkovic) [reply packages/client/src/components/TransferToken/components/CoinTypeDropDown.js]: I'm not sure, I'll revert it!

File: packages/client/src/modals/Transaction/BannerInfo.js
Comment (mannynotfound) [reply packages/client/src/modals/Transaction/BannerInfo.js:25]: @KsenijaZivkovic this should work for both I think:

`const imageURI = nftMeta?.imageURI ?? nftMeta?.thumbnail;`

**Review by mannynotfound — APPROVED** [review]

File: packages/client/src/components/SendTokens/steps/DepositTokenForm.js
Comment (linxiao-li) [reply packages/client/src/components/SendTokens/steps/DepositTokenForm.js]: can we do something like this?
```
useEffect(()=>{
    if(isEmpty(vaults)){
         getTreasuryVaults();
    }
    updateUserBalance();
}, [userAddress])
```

Comment (KsenijaZivkovic) [reply packages/client/src/components/SendTokens/steps/DepositTokenForm.js]: Updated it like this:
```
  useEffect(() => {
    if (!userAddress) {
      return;
    }
    if (isEmpty(vaults)) {
      getVaults();
    } else {
      updateUserBalance();
    }
  }, [userAddress, vaults]);
  ```

**Review by linxiao-li — APPROVED** [review]
LGTM

**Pull request merged**
